src/shared/reboot-util.c

```diff
@@ -107,3 +124,67 @@ int shall_restore_state(void) {
 
         return r > 0 ? ret : true;
 }
+
+static int xen_kexec_loaded(void) {
+#if HAVE_XENCTRL
+        size_t size;
+        _cleanup_close_ int privcmd_fd = -1, buf_fd = -1;
+        void *buffer;
+        int r;
+
+        if (access("/proc/xen", F_OK) < 0)
+                return -EOPNOTSUPP;
+
+        size = page_size();
+        if (sizeof(xen_kexec_status_t) > size)
+                return log_debug_errno(SYNTHETIC_ERRNO(EINVAL), "page_size is too small for hypercall");
+
+        privcmd_fd = open("/dev/xen/privcmd", O_RDWR|O_CLOEXEC);
+        if (privcmd_fd < 0)
+                return log_debug_errno(errno, "Cannot access /dev/xen/privcmd: %m");
+
+        buf_fd = open("/dev/xen/hypercall", O_RDWR|O_CLOEXEC);
+        if (buf_fd < 0)
+                return log_debug_errno(errno, "Cannot access /dev/xen/hypercall: %m");
+
+        buffer = mmap(NULL, size, PROT_READ|PROT_WRITE, MAP_SHARED, buf_fd, 0);
+        if (buffer == MAP_FAILED)
+                return log_error_errno(errno, "Cannot allocate buffer for hypercall: %m");
```

Comment:
downgrade to log_debug_errno() please

Reply:
Why? this is really not supposed to happen, and a sign that the system is behaving oddly.

Reply:
we have this rule in systemd: functions either log about all their errors or about none (in which case the caller may decide to log). Whenever you write a function, figure out which kind of function yours is: a logging one (typically the more main-program-like stuff) or a non-logging one (typically library-like stuff). 

Via this rule we ensure that every error gets logged exactly once, and not accidentally eaten up entirely, or logged about twice (both of which are confusing to the user).

(note that LOG_DEBUG logging does not count as "logging" in this context, debug logging is always OK.)

https://systemd.io/CODING_STYLE/#error-handling

Your function so far looks like a non-logging one (and I think that's the right choice, given this is a library-like thing). Hence it should not log above LOG_DEBUG, on any code path.